### PR TITLE
[None][fix] use prompt lookahead for MTP Eagle chunked prefill

### PR DESCRIPTION
**Pull request by yizhang-nv**
## Description

MTP Eagle left-shifts each context chunk before the first draft-layer forward. For a non-final chunk, the shifted tail must be the immediate next prompt token. It previously used the target model's sampled token, so linear and dynamic-tree MTP Eagle could persist draft-layer KV that did not match canonical full-prompt prefill.

This change:

- stages exactly one immediate prompt lookahead token per active context request in a fixed-address `SpecMetadata` tensor with shape `[max_num_requests]`; `-1` marks final chunks with no valid lookahead;
- derives the token from the request's context cursor, independently of KV cache manager V1 or V2;
- uses the prompt token for non-final context rows in linear, dynamic-tree, and external shared-target-KV MTP Eagle, while final chunks fall back to the sampled token and generation rows remain unchanged; and
- leaves vanilla MTP, cache keys, and V2 draft reuse unchanged for follow-up work.

The external shared-target-KV path does not persist independent draft KV, but it uses the same token/hidden-state/position contract so all MTP-Eagle rollout paths now start from the correctly aligned token.

## Acceptance Measurements

Measured on one B200 with `DeepSeek-V3-Lite/bf16`, TP=1, MTP Eagle one-model, max draft length 1, chunked prefill enabled, prompt length 512, chunk size 128, and output length 128. The workload used three prompts sourced from the TensorRT-LLM documentation and two runs per configuration.

| Configuration | Accepted / drafted tokens | Acceptance rate | Acceptance length |
| --- | ---: | ---: | ---: |
| Before | 338 / 430 | 78.60% | 1.7860 |
| After | 339 / 429 | 79.02% | 1.7902 |

The aggregate delta is +0.42 percentage points in acceptance rate and +0.0042 in acceptance length. Per-run acceptance-rate ranges overlap (before: 76.96%-80.28%; after: 78.60%-79.44%), so this should be treated as neutral/small with no observed regression, rather than as a statistically meaningful improvement.

## Test Coverage

- `pytest -s -q tests/unittest/_torch/executor/test_pytorch_model_engine.py tests/unittest/_torch/speculative/test_eagle3.py tests/unittest/_torch/speculative/hw_agnostic/test_mtp.py`
  - 128 passed, 22 skipped, 26 subtests passed
- `pre-commit run --files` on all seven changed files
  - all hooks passed

The focused tests cover prompt-boundary extraction, token ID `0`, sentinel fallback and stale-value clearing, CUDA graph metadata allocation, linear and dynamic-tree context input, and external shared-target-KV context/generation alignment.

## PR Checklist

- [x] The fix is independent of KV cache manager V1/V2.
- [x] Linear, dynamic-tree, and external shared-target-KV MTP-Eagle paths are covered.
- [x] Vanilla MTP and V2 draft reuse are intentionally unchanged.
- [x] No public API, dependency, ownership, documentation, or architecture-diagram change is introduced.

## Dev Engineer Review

- The implementation stores one immediate prompt lookahead token per active context request.
- `-1` marks final chunks and preserves sampled-token fallback behavior.
- The implementation is independent of KV cache manager V1 and V2.
- Linear, dynamic-tree, and shared-target-KV MTP Eagle paths use the lookahead token.
- Vanilla MTP, cache keys, and V2 draft reuse remain unchanged.
- The `SpecMetadata` API now uses a flat one-token-per-request buffer.
- No configuration or test-list files changed.
- Focused tests passed: 128 passed, 22 skipped, and 26 subtests passed.
- Pre-commit checks passed for all seven changed files.
- Reported measurements improved acceptance rate from 78.60% to 79.02%, with overlapping per-run ranges and no observed regression.

## QA Engineer Review

- Modified test files:
  - `test_pytorch_model_engine.py`: updated prompt-boundary and overlap-mock coverage.
  - `test_eagle3.py`: updated one-token lookahead, sentinel, CUDA-graph, draft-input, and dynamic-tree coverage.
  - `test_mtp.py`: updated shared-KV draft-input coverage for multiple context requests.
- The tests cover non-final lookahead tokens, final-chunk sentinel fallback, accepted tokens, draft IDs, recurrent states, and positions.
- No corresponding `test-db/` or `qa/` test-list changes were provided.
- The modified test functions are not mapped in the provided test-list data.
- Verdict: needs follow-up. CBTS coverage mapping is unavailable.

### CODE REVIEW

**Review by reasonsolo — APPROVED** [review]

**Review by zhaoyangwang-nvidia — COMMENTED** [review]
Could you add measured acceptance-rate / acceptance-length numbers, before and after this change, to the description — along with the prompt length and chunk size used?

File: tensorrt_llm/_torch/pyexecutor/model_engine.py [hunk 1]
```diff
@@ -6585,6 +6605,12 @@ def _prepare_tp_inputs_no_cache(
                 scheduled_requests.generation_requests)
             spec_metadata.num_tokens = num_tokens
             spec_metadata.seq_lens = sequence_lengths
+            if spec_metadata.context_prompt_lookahead_tokens is not None:
+                # No-cache context inputs contain the complete prompt, so
+                # there is never a valid token beyond the current chunk.
+                spec_metadata.populate_context_prompt_lookahead(
```
Comment (zhaoyangwang-nvidia) [inline tensorrt_llm/_torch/pyexecutor/model_engine.py]: `_set_up_spec_metadata`'s `no_cache` branch constructs a fresh `SpecMetadata` on every call rather than reusing `self.spec_metadata`, so the buffer here was just filled with `INVALID_PROMPT_LOOKAHEAD_TOKEN` by `allocate_context_prompt_lookahead` and this call always writes `-1` over `-1`. Is there a path where a stale value could actually reach here? If not, I'd drop the block rather than leave a no-op that implies one exists.

File: tensorrt_llm/_torch/speculative/interface.py [hunk 2]
```diff
@@ -661,6 +670,38 @@ def batch_uses_penalty(self, value: bool) -> None:
...
+    def populate_context_prompt_lookahead(self,
+                                          lookahead_tokens: List[int]) -> None:
+        """Populate immediate prompt lookahead for this batch's context rows."""
+        if self.context_prompt_lookahead_tokens is None:
+            return
+
+        num_contexts = len(lookahead_tokens)
+        if num_contexts == 0:
+            return
+        if num_contexts > self.context_prompt_lookahead_tokens.shape[0]:
```
Comment (zhaoyangwang-nvidia) [inline tensorrt_llm/_torch/speculative/interface.py]: This branch looks unreachable: on the normal path `max_num_requests` is the engine batch size and the scheduler bounds the batch by it, and under CUDA graph the metadata copy shrinks `max_num_requests` but `num_contexts` is 0 (padded batches set it to 0, and the prefill-graph path promotes final-chunk contexts into `generation_requests`). Since this is a scheduler invariant rather than user input, and it runs on every forward with a context batch, an `assert` — strippable under `python -O`, and consistent with the `batch_indices_cuda` shape check in `__post_init__` — reads better than a `ValueError`.

File: tensorrt_llm/_torch/speculative/eagle3.py [hunk 3]
```diff
@@ -1356,8 +1361,14 @@ def prepare_1st_drafter_inputs(
...
-            input_ids, attn_metadata.num_ctx_tokens, spec_metadata.gather_ids,
-            accepted_tokens, num_contexts)
+            input_ids,
+            attn_metadata.num_ctx_tokens,
+            spec_metadata.gather_ids,
+            accepted_tokens,
+            num_contexts,
+            prompt_lookahead_tokens=(
+                spec_metadata.context_prompt_lookahead_tokens
+                if self.is_mtp_eagle else None))
```
Comment (zhaoyangwang-nvidia) [inline tensorrt_llm/_torch/speculative/eagle3.py]: `context_prompt_lookahead_tokens` is only allocated when `spec_dec_mode.is_mtp_eagle_one_model()`, and the metadata is built with `spec_dec_mode=spec_config.spec_dec_mode` (`utils.py:380`) — the same source as `self.is_mtp_eagle`. So the field is already `None` outside MTP Eagle and this ternary is redundant. Dropping it would also make this call site consistent with `_forward_external_shared_target_kv_draft_loop`, which passes the same field unguarded into a parameter that is dereferenced without a `None` check.

File: tensorrt_llm/_torch/speculative/eagle3.py [hunk 4]
```diff
@@ -1394,6 +1406,13 @@ def _prepare_shared_kv_draft_inputs(
             num_accepted_tokens[num_contexts:] - 1)
         draft_input_ids = accepted_tokens[batch_indices,
                                           num_accepted_tokens - 1]
+        context_lookahead = prompt_lookahead_tokens[:num_contexts]
+        draft_input_ids[:num_contexts].copy_(
```
Comment (zhaoyangwang-nvidia) [inline tensorrt_llm/_torch/speculative/eagle3.py]: `draft_input_ids` comes from two-tensor advanced indexing, so it is already a fresh tensor — `draft_input_ids[:num_contexts] = torch.where(...)` avoids the round-trip through `.copy_()`. Separately, the new `prompt_lookahead_tokens` parameter is undocumented: the `_prepare_shared_kv_draft_inputs` docstring on line 1400 was shortened in this PR and has no `Args` section, so please add one there describing the shape and the sentinel meaning — mirroring the entry added for the same parameter in `interface.py:2663`.

**Review by zhaoyangwang-nvidia — APPROVED** [review]

**Review by Shixiaowei02 — APPROVED** [review]

**Comment by yizhang-nv** [comment]
@zhaoyangwang-nvidia Thanks for the review. I added the measured acceptance-rate and acceptance-length results, including prompt length and chunk size, to the PR description.

On one B200 with DeepSeek-V3-Lite BF16, TP=1, MTP Eagle one-model, max draft length 1, prompt length 512, chunk size 128, and output length 128, using 3 documentation prompts and 2 runs per configuration:

- before: 338/430 accepted draft tokens, AR 78.60%, acceptance length 1.7860
- after: 339/429 accepted draft tokens, AR 79.02%, acceptance length 1.7902

That is +0.42 pp AR and +0.0042 acceptance length. The per-run AR ranges overlap (76.96%-80.28% before and 78.60%-79.44% after), so I described the impact as neutral/small with no observed regression rather than claiming a statistically meaningful improvement.

I also addressed the inline cleanup suggestions, rebased onto the latest origin/main, and reran the related suites: 128 passed, 22 skipped, and 26 subtests passed.
